src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php

```diff
@@ -110,9 +110,9 @@ public function setUser($user)
 
         if ($changed) {
             $this->setAuthenticated(false);
+        } else {
+            $this->user = $user;
         }
-
-        $this->user = $user;
```

Comment:
Well as I stated at the top of my initial comment, this is the point that is actually the most problematic. And I need input frim somebody that knows really well the security component and it's usage. 

This setter as it stands does too much on it's own, and should be refactored. Most likely separating the logic for marking a token as not authenticated and doing a check elsewhere before setting the token.

The logout fix will not work if we overwrite the user in the token with the refreshed user.